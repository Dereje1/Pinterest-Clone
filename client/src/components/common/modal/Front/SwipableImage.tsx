// Import necessary modules and components
import React, { useEffect, useState } from 'react';
import { useSwipeable } from 'react-swipeable';
import CardMedia from '@mui/material/CardMedia';
import ImageListItemBar from '@mui/material/ImageListItemBar';
import IconButton from '@mui/material/IconButton';
import NavigateNextIcon from '@mui/icons-material/NavigateNext';
import NavigateBeforetIcon from '@mui/icons-material/NavigateBefore';
import { zoomedImageInfoType, imageMetadataType } from '../../../../interfaces';
import LoadingImage from '../../../../assets/giphy.gif';

// Define the props for the SwipableImage component
interface SwipableImageProps {
  zoomInfo: zoomedImageInfoType,
  onSlidePin: (newIndex: number) => void
  onSetImageMetaData: (metaData: imageMetadataType) => void
}

// Configuration for the swipeable component
const config = {
  delta: 10, // min distance(px) before a swipe starts
  preventScrollOnSwipe: false, // prevents scroll during swipe
  trackTouch: true, // track touch input
  trackMouse: true, // track mouse input
  rotationAngle: 0, // set a rotation angle
  swipeDuration: Infinity, // allowable duration of a swipe (ms)
  touchEventOptions: { passive: true }, // options for touch listeners
};

// Define the type for the slide direction
type slideType = 'left' | 'right' | false

// Define the SwipableImage component
function SwipableImage({ zoomInfo, onSlidePin, onSetImageMetaData }: SwipableImageProps) {
  const { pin: pinInformation, parentDivStyle } = zoomInfo;

  // Define the state variables
  const [slideImage, setSlideImage] = useState(false as slideType);
  const [showImageListItemBar, setShowImageListItemBar] = useState(false);
  const [isLoading, setIsLoading] = useState(false); // State for loading status

  // Effect hook to handle slide image state changes
  useEffect(
    () => {
      if (slideImage === 'left') {
        setIsLoading(true);
        onSlidePin(zoomInfo.loadedIndex - 1);
      }

      if (slideImage === 'right') {
        setIsLoading(true);
        onSlidePin(zoomInfo.loadedIndex + 1);
      }

      return () => {
        setSlideImage(false);
      };
    },
    [slideImage],
  );

  // Function to handle image load events
  const onPinLoad = (e: React.SyntheticEvent<HTMLImageElement>) => {
    const target = e.target as HTMLImageElement;
    const { naturalWidth, naturalHeight } = target;
    onSetImageMetaData({ naturalWidth, naturalHeight });
    setIsLoading(false);
  };

  // Define the swipe handlers
  const handlers = useSwipeable({
    onSwipedLeft: () => setSlideImage('left'),
    onSwipedRight: () => setSlideImage('right'),
    ...config,
  });

  // Return the component JSX
  return (
    <div
      onMouseOver={() => setShowImageListItemBar(true)}
      onMouseLeave={() => setShowImageListItemBar(false)}
      onFocus={() => ({})}
    >
      <CardMedia
        component="img"
        image={isLoading ? LoadingImage : pinInformation.imgLink}
        id="pin-zoom"
        sx={{
          width: parentDivStyle.imgWidth,
          height: parentDivStyle.imgHeight,
          marginLeft: 'auto',
          marginRight: 'auto',
          objectFit: 'contain',
          border: parentDivStyle.border,
        }}
        onLoad={onPinLoad}
        {...handlers}
      />
      {
        showImageListItemBar && (
          <ImageListItemBar
            position="bottom"
            sx={
              {
                background: 'rgba(0, 0, 0, 0.5)',
                '& .MuiImageListItemBar-titleWrap': {
                  flexGrow: 0,
                  display: 'none',
                },
                '& .MuiImageListItemBar-actionIcon': {
                  flexGrow: 1,
                },
              }
            }
            actionIcon={(
              <div id="actionbuttons" style={{ display: 'flex', justifyContent: 'space-around', width: '100%' }}>
                <IconButton
                  sx={{ color: 'rgba(255, 255, 255, 1)' }}
                  aria-label={`info about ${pinInformation.imgDescription}`}
                  onClick={() => setSlideImage('left')}
                  onMouseDown={(e) => e.preventDefault()}
                >
                  <NavigateBeforetIcon fontSize="large" />
                </IconButton>
                <IconButton
                  sx={{ color: 'rgba(255, 255, 255, 1)' }}
                  aria-label={`info about ${pinInformation.imgDescription}`}
                  onClick={() => setSlideImage('right')}
                  onMouseDown={(e) => e.preventDefault()}
                >
                  <NavigateNextIcon fontSize="large" />
                </IconButton>
              </div>
            )}
          />
        )
      }

    </div>
  );
}

export default SwipableImage;
