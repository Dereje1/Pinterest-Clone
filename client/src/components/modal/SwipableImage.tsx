import React from 'react';
import { useSwipeable } from 'react-swipeable';
import CardMedia from '@mui/material/CardMedia';
import ImageListItemBar from '@mui/material/ImageListItemBar';
import IconButton from '@mui/material/IconButton';
import NavigateNextIcon from '@mui/icons-material/NavigateNext';
import NavigateBeforetIcon from '@mui/icons-material/NavigateBefore';
import { zoomedImageInfoType } from '../../interfaces';

interface SwipableImageProps {
    zoomInfo: zoomedImageInfoType,
    onSwipe: (newIndex: number) => void
}

const config = {
  delta: 10, // min distance(px) before a swipe starts. *See Notes*
  preventScrollOnSwipe: false, // prevents scroll during swipe (*See Details*)
  trackTouch: true, // track touch input
  trackMouse: true, // track mouse input
  rotationAngle: 0, // set a rotation angle
  swipeDuration: Infinity, // allowable duration of a swipe (ms). *See Notes*
  touchEventOptions: { passive: true }, // options for touch listeners (*See Details*)
};

function SwipableImage({ zoomInfo, onSwipe }: SwipableImageProps) {
  const { pin: pinInformation, parentDivStyle } = zoomInfo;
  const handlers = useSwipeable({
    onSwipedLeft: () => onSwipe(zoomInfo.loadedIndex + 1),
    onSwipedRight: () => onSwipe(zoomInfo.loadedIndex - 1),
    ...config,
  });

  return (
    <>
      <CardMedia
        component="img"
        image={pinInformation.imgLink}
        id="pin-zoom"
        sx={{
          width: parentDivStyle.imgWidth,
          marginLeft: 'auto',
          marginRight: 'auto',
        }}
        {...handlers}
      />
      <ImageListItemBar
        position="bottom"
        sx={
          {
            background: 'transparent',
            '& .MuiImageListItemBar-titleWrap': {
              flexGrow: 0,
              display: 'none',
            },
            '& .MuiImageListItemBar-actionIcon': {
              flexGrow: 1,
            },
          }
        }
        actionIcon={(
          <div style={{ display: 'flex', justifyContent: 'space-around', width: '100%' }}>
            <IconButton
              sx={{ color: 'rgba(255, 255, 255, 1)' }}
              aria-label={`info about ${pinInformation.imgDescription}`}
              onClick={() => onSwipe(zoomInfo.loadedIndex - 1)}
              onMouseDown={(e) => e.preventDefault()}
            >
              <NavigateBeforetIcon />
            </IconButton>
            <IconButton
              sx={{ color: 'rgba(255, 255, 255, 1)' }}
              aria-label={`info about ${pinInformation.imgDescription}`}
              onClick={() => onSwipe(zoomInfo.loadedIndex + 1)}
              onMouseDown={(e) => e.preventDefault()}
            >
              <NavigateNextIcon />
            </IconButton>
          </div>
        )}
      />
    </>
  );
}

export default SwipableImage;
