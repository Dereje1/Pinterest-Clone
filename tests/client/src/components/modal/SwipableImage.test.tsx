/**
 * @jest-environment jsdom
 */
import React from 'react';
import { EnzymePropSelector, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import SwipableImage from '../../../../../client/src/components/modal/SwipableImage';
import { pinsStub } from '../../../stub';

const props: React.ComponentProps<typeof SwipableImage> = {
  zoomInfo: {
    pin: pinsStub[0],
    parentDivStyle: {
      top: 10,
      imgWidth: 622.5,
      parentWidth: 622.5,
      width: '90%',
      isNoFit: false,
    },
    loadedIndex: 0,
  },
  onSwipe: jest.fn(),
};

test('will render', () => {
  const wrapper = shallow(<SwipableImage {...props} />);
  const img:EnzymePropSelector = wrapper.find('ForwardRef(CardMedia)');
  expect(toJson(wrapper)).toMatchSnapshot();
  img.props().onMouseDown({ });
});

test('will handle manual forward and backward clicks', () => {
  const wrapper = shallow(<SwipableImage {...props} />);
  const listItemBar:EnzymePropSelector = wrapper.find('ForwardRef(ImageListItemBar)');
  const back = listItemBar.props().actionIcon.props.children[0];
  const next = listItemBar.props().actionIcon.props.children[1];
  back.props.onClick();
  back.props.onMouseDown({ preventDefault: jest.fn() });
  next.props.onClick();
  next.props.onMouseDown({ preventDefault: jest.fn() });
  expect(props.onSwipe).toHaveBeenCalledWith(-1);
  expect(props.onSwipe).toHaveBeenCalledWith(1);
});
