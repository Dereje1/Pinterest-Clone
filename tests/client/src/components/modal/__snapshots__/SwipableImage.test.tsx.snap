// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will render 1`] = `
<Fragment>
  <ForwardRef(CardMedia)
    component="img"
    id="pin-zoom"
    image="imgLink id-1"
    onMouseDown={[Function]}
    sx={
      {
        "marginLeft": "auto",
        "marginRight": "auto",
        "width": 622.5,
      }
    }
  />
  <ForwardRef(ImageListItemBar)
    actionIcon={
      <div
        style={
          {
            "display": "flex",
            "justifyContent": "space-around",
            "width": "100%",
          }
        }
      >
        <ForwardRef(IconButton)
          aria-label="info about imgDescription id-1"
          onClick={[Function]}
          onMouseDown={[Function]}
          sx={
            {
              "color": "rgba(255, 255, 255, 1)",
            }
          }
        >
          <Memo />
        </ForwardRef(IconButton)>
        <ForwardRef(IconButton)
          aria-label="info about imgDescription id-1"
          onClick={[Function]}
          onMouseDown={[Function]}
          sx={
            {
              "color": "rgba(255, 255, 255, 1)",
            }
          }
        >
          <Memo />
        </ForwardRef(IconButton)>
      </div>
    }
    position="bottom"
    sx={
      {
        "& .MuiImageListItemBar-actionIcon": {
          "flexGrow": 1,
        },
        "& .MuiImageListItemBar-titleWrap": {
          "display": "none",
          "flexGrow": 0,
        },
        "background": "transparent",
      }
    }
  />
</Fragment>
`;
